.github/workflows/test-go.yml

```diff
@@ -108,13 +113,29 @@ jobs:
       - name: List cached results
         id: list-cached-results
         run: ls -lhR test-results/go-test
-      - name: Build matrix excluding binary and integration tests
+      - name: Build matrix excluding binary, integration, and testonly tests
         id: build-non-binary
+        if: ${{ !inputs.testonly }}
         env:
           GOPRIVATE: github.com/hashicorp/*
         run: |
+          # testonly tests need additional build tag though let's exclude them anyway for clarity
+          (
+            go list ./... | grep -v "_binary" | grep -v "vault/integ" | grep -v "testonly" | gotestsum tool ci-matrix --debug \
+              --partitions "${{ inputs.total-runners }}" \
+              --timing-files 'test-results/go-test/*.json' > matrix.json
+          )
+      - name: Build matrix for tests tagged with testonly
+        if: ${{ inputs.testonly }}
+        env:
+          GOPRIVATE: github.com/hashicorp/*
+        run: |
+          set -exo pipefail
+          
+          # testonly tagged tests need an additional tag to be included
+          # also running some extra tests for sanity checking with the testonly build tag
           (
-            go list ./... | grep -v "_binary" | grep -v "vault/integ" | gotestsum tool ci-matrix --debug \
+            go list -tags=testonly ./... | grep -E -i 'external_tests/kv|external_tests/token|testonly|external_tests/replication-perf|^github.com/hashicorp/vault/vault$' | gotestsum tool ci-matrix --debug \
```

Comment:
Extension of my other comments: I don't think the `-tags` flag here does what we expect it to, since `go list` is a package listing mechanism and `-tags` is a build mechanism:

```
  ✭ ◼ [remotes/origin/HEAD] ~/s/vault >  go list -tags=testonly ./... >  testonly.out
  ✭ ◼ [remotes/origin/HEAD] ~/s/vault >  go list ./... > notestonly.out
  ✭ ◼ [remotes/origin/HEAD] ~/s/vault >  git diff --no-index testonly.out notestonly.out
  ✭ ◼ [remotes/origin/HEAD] ~/s/vault >  
  ```

Reply:
I am about to update my PR in ENT, and you will see the difference once I added the test package.

Reply:
Sounds good. I think the point still stands that `-tags=` here doesn't have any build-tag awareness, so minimally we can just omit it. I do think we should put a bit of thought into how we want to filter these packages though. Maybe there's a simpler solution that doesn't require refactoring.

Reply:
I stand corrected! Looks like if a package only has `testonly`-tagged files, you need the `-tags` flag to list them. Sorry about that @hghaf099!